File: apps/api/app/schemas/brand_profiles.py
from datetime import datetime

from pydantic import BaseModel, Field


class BrandProfileRecord(BaseModel):
    profile_id: str
    organization_id: str = "org-default"
    owner_user_id: str = "system"
    brand_name: str
    trademark_classes: list[str] = Field(default_factory=list)
    trademark_numbers: list[str] = Field(default_factory=list)
    confusable_terms: list[str] = Field(default_factory=list)
    protection_keywords: list[str] = Field(default_factory=list)
    created_at: datetime
    updated_at: datetime


class BrandProfileCreateRequest(BaseModel):
    brand_name: str = Field(..., min_length=1)
    trademark_classes: list[str] = Field(default_factory=list)
    trademark_numbers: list[str] = Field(default_factory=list)
    confusable_terms: list[str] = Field(default_factory=list)
    protection_keywords: list[str] = Field(default_factory=list)


class BrandProfileUpdateRequest(BaseModel):
    brand_name: str | None = None
    trademark_classes: list[str] | None = None
    trademark_numbers: list[str] | None = None
    confusable_terms: list[str] | None = None
    protection_keywords: list[str] | None = None


class BrandProfileListResponse(BaseModel):
    total: int
    items: list[BrandProfileRecord]


class SuggestConfusableResponse(BaseModel):
    brand_name: str
    suggestions: list[str]

File: apps/web/lib/brand-profiles.ts
import { fetchJsonOrUndefined, buildApiUrl } from "@/lib/api";
import { getApiAuthToken } from "@/lib/env";
import { buildApiErrorNote, type ListFetchResult } from "@/lib/data-source";

export type BrandProfile = {
  profile_id: string;
  organization_id: string;
  brand_name: string;
  trademark_classes: string[];
  trademark_numbers: string[];
  confusable_terms: string[];
  protection_keywords: string[];
  created_at: string;
  updated_at: string;
};

export type BrandProfileCreatePayload = {
  brand_name: string;
  trademark_classes: string[];
  trademark_numbers: string[];
  confusable_terms: string[];
  protection_keywords: string[];
};

type ApiBrandProfileList = {
  total?: number;
  items?: BrandProfile[];
};

export async function fetchBrandProfiles(): Promise<ListFetchResult<BrandProfile>> {
  const data = await fetchJsonOrUndefined<ApiBrandProfileList>("brand-profiles");
  if (!data) {
    return { items: [], source: "error", note: buildApiErrorNote("brand-profiles") };
  }
  return { items: data.items ?? [], source: "api", total: data.total };
}

async function mutateProxyJson(path: string, method: string, body?: unknown): Promise<Response> {
  const url = buildApiUrl(path);
  const token = getApiAuthToken().trim();
  return fetch(url, {
    method,
    headers: {
      "content-type": "application/json",
      ...(token ? { Authorization: `Bearer ${token}` } : {}),
    },
    body: body !== undefined ? JSON.stringify(body) : undefined,
  });
}

export async function createBrandProfile(payload: BrandProfileCreatePayload): Promise<BrandProfile | null> {
  const res = await mutateProxyJson("brand-profiles", "POST", payload);
  if (!res.ok) return null;
  return res.json();
}

export async function updateBrandProfile(
  profileId: string,
  payload: Partial<BrandProfileCreatePayload>
): Promise<BrandProfile | null> {
  const res = await mutateProxyJson(`brand-profiles/${profileId}`, "PUT", payload);
  if (!res.ok) return null;
  return res.json();
}

export async function deleteBrandProfile(profileId: string): Promise<boolean> {
  const res = await mutateProxyJson(`brand-profiles/${profileId}`, "DELETE");
  return res.ok;
}

export async function suggestConfusable(profileId: string): Promise<string[]> {
  const res = await mutateProxyJson(`brand-profiles/${profileId}/suggest-confusable`, "POST");
  if (!res.ok) return [];
  const data = await res.json();
  return data.suggestions ?? [];
}
